## String Methods:

[__capitalize()__](#capitalize) --- _similar_ --- [__lower()__](#lower) --- _similar_ --- [__upper()__](#upper)  --- _similar_ ---  [__swapcase()__](#swapcase)  
[casefold()](#casefold)  
[__center()__](#center)  
[__count()__](#count)  
[encode()](#encode)  
[__endswith()__](#endswith) ---similar--- [__startswith()__](#startswith)  
[expandtabs()](#expandtabs)  
[__find()__](#find) ---similar--- [**rfind()**](#rfind) ---similar--- [**index()**](#index)  
1. [format()](#format)
10. [format_map()](#format_map)
12. [isalnum()](#isalnum)
13. [isalpha()](#isalpha)
14. [isdecimal()](#isdecimal)
15. [isdigit()](#isdigit)
16. [isidentifier()](#isidentifier)
17. [islower()](#islower)
18. [isnumeric()](#isnumeric)
19. [isprintable()](#isprintable)
20. [isspace()](#isspace)
21. [istitle()](#istitle)
22. [isupper()](#isupper)
23. [join()](#join)
24. [ljust()](#ljust)
26. [lstrip()](#lstrip)
27. [maketrans()](#maketrans)
28. [partition()](#partition)
29. [replace()](#replace)
31. [rindex()](#rindex)
32. [rjust()](#rjust)
33. [rpartition()](#rpartition)
34. [rsplit()](#rsplit)
35. [rstrip()](#rstrip)
36. [split()](#split)
37. [splitlines()](#splitlines)
39. [strip()](#strip)
41. [title()](#title)
42. [translate()](#translate)

44. [zfill()](#zfill)

## <a name="capitalize"></a>1. capitalize()

__Definition__: converts first character of a string to uppercase letter and lowercases all other characters.   

__Syntax__:  
    
    string.capitalize()  
    
__Parameters__:  
    - doesn't take any parameter  
__Returns__:  
    - returns a string with the first letter capitalized and all other characters lowercased  
    - It doesn't modify the original string  

__Similar:__ [lower()](#lower) / [upper()](#upper) / [swapcase()](#swapcase)  

In [24]:
a_str = "python is NIce"
new_str = a_str.capitalize()

print('Old string: ', a_str)
print('New string: ', new_str)

Old string:  python is NIce
New string:  Python is nice


In [3]:
a_str = " python is NIce"
new_str = a_str.capitalize()

print('Old string: ', a_str)
print('New string: ', new_str)

Old string:   python is NIce
New string:   python is nice


In [4]:
a_str = "+python is NIce"
new_str = a_str.capitalize()

print('Old string: ', a_str)
print('New string: ', new_str)

Old string:  +python is NIce
New string:  +python is nice


In [5]:
a_str = "1pYthon is NIce"
new_str = a_str.capitalize()

print('Old string: ', a_str)
print('New string: ', new_str)

Old string:  1pYthon is NIce
New string:  1python is nice


## <a name="casefold"></a>2. casefold()

__Definition__: an aggressive lower() method which converts strings to case folded strings for caseless matching.   

- removes all case distinctions present in a string.  
- used for caseless matching, i.e. ignores cases when comparing.  

__For example__:  
the German lowercase letter ß is equivalent to ss.  
ß is already lowercase.  
lower() does nothing to it.  
casefold() converts it to ss.  

__Syntax__:  
    
    string.casefold()  
    
__Parameters__:  
    - doesn't take any parameter  
__Returns__:  
    - the case folded string  

In [6]:
string = "PYTHON IS AWESOME"

# print lowercase string
print("Lowercase string:", string.casefold())

Lowercase string: python is awesome


In [7]:
firstString = "der Fluß"
secondString = "der Fluss"

# ß is equivalent to ss
if firstString.casefold() == secondString.casefold():
    print('The strings are equal.')
else:
    print('The strings are not equal.')

The strings are equal.


## <a name="center"></a>3. center()

__Definition__: returns a string which is padded with the specified character.  

__Syntax__:  
    
    string.center(width[, fillchar])  
    
__Parameters__:  
    - width: length of the string with padded characters  
    - fillchar (optional): padding character
               ! if not provided - space as default
__Returns__:  
    - returns a string padded with specified fillchar.  
    - It doesn't modify the original string.  

In [2]:
string = "Python is awesome"
new_string = string.center(24)
print("Centered String: ", new_string)

Centered String:     Python is awesome    


In [3]:
string = "Python is awesome"
new_string = string.center(24, '*')
print("Centered String: ", new_string)

Centered String:  ***Python is awesome****


## <a name="count"></a>4. count()

__Definition__: returns the number of occurrences of a substring in the given string.  

__Syntax__:  
    
    string.count(substring, start=..., end=...)  
    
__Parameters__:  
    - substring: string whose count is to be found  
    - start (optional): starting index within the string where search starts  
    - end (optional): ending index within the string where search starts  
__Returns__:  
    - the number of occurrences of the substring in the given string  

In [4]:
string = "Python is awesome, isn't it?"
substring = "is"
count = string.count(substring)
# print count
print("The count is:", count)

The count is: 2


In [5]:
string = "Python is awesome, isn't it?"
substring = "i"

# count after first 'i' and before the last 'i'
count = string.count(substring, 8, 25)

# print count
print("The count is:", count)

The count is: 1


## <a name="encode"></a>5. encode()

__Definition__: returns encoded version of the given string.  
! Strings in Python - stored in as Unicode - each character = represented by a code point  
! the sequence of code points - converted in bytes = __Encoding__  
! popular encodings: utf-8, ascii  
! encode() - converts unicode strings into any encodings supported

__Syntax__:  
    
    string.encode(encoding='UTF-8', errors='strinct')  
    
__Parameters__:  
    - does not require any
    - if failure => UnicodeDecodeError  
    - encoding (optional): encoding type a string has to be encoded to  
    - errors (optional): response when encoding fails.  
        - strinct: default response => UnicodeDecodeError  
        - ignore: ignores the unencodable unicode from the result  
        - replace: replaces the unencodable unicode to ?  
        - xmlcharrefreplace: inserts XML character reference instead of unencodable unicode  
        - backslashreplace: inserts a \uNNNN escape sequence instead of unencodable unicode  
        - namereplace: inserts a \N{...} escape sequence instead of unencodable unicode  
__Returns__:  
    - utf-8 encoded version of the string  

In [6]:
# unicode string
string = 'pythön!'

# print string
print('The string is:', string)

# default encoding to utf-8
string_utf = string.encode()

# print result
print('The encoded version is:', string_utf)

The string is: pythön!
The encoded version is: b'pyth\xc3\xb6n!'


In [7]:
# unicode string
string = 'pythön!'

# print string
print('The string is:', string)

# ignore error
print('The encoded version (with ignore) is:', string.encode("ascii", "ignore"))

# replace error
print('The encoded version (with replace) is:', string.encode("ascii", "replace"))

The string is: pythön!
The encoded version (with ignore) is: b'pythn!'
The encoded version (with replace) is: b'pyth?n!'


## <a name="endswith"></a>6. endswith()

__Definition__: returns True if a string ends with the specified suffix. If not, it returns False.  

__Syntax__:  
    
    string.endswith(suffix[start,[ end]])  
    
__Parameters__:  
    - suffix: String or tuple of suffixes to be checked  
    - start (optional): Beginning position where suffix is to be checked within the string  
    - end (optional): nding position where suffix is to be checked within the string  
__Returns__:  
    - boolean:  
        - True - if string ends with suffix  
        - False - if string doesn't ends with suffix  

__Similar:__ [startswith()](#startswith)

In [9]:
text = "Python is easy to learn."

result = text.endswith('to learn')
# returns False
print(result)

result = text.endswith('to learn.')
# returns True
print(result)

result = text.endswith('Python is easy to learn.')
# returns True
print(result)

False
True
True


In [11]:
text = "Python programming is easy to learn."

# start parameter: 7
# "programming is easy to learn." string is searched
result = text.endswith('learn.', 7)
print(result)

# Both start and end is provided
# start: 7, end: 26
# "programming is easy" string is searched

result = text.endswith('is', 7, 26)
# Returns False
print(result)

result = text.endswith('easy', 7, 26)
# returns True
print(result)

True
False
True


### Passing Tuple to endswith()

It's possible to pass a tuple suffixes to the endswith() method in Python.  
If the string ends with any item of the tuple, endswith() returns True.  
If not, it returns False

In [12]:
text = "programming is easy"
result = text.endswith(('programming', 'python'))

# prints False
print(result)

result = text.endswith(('python', 'easy', 'java'))

#prints True
print(result)

# With start and end parameter
# 'programming is' string is checked
result = text.endswith(('is', 'an'), 0, 14)

# prints True
print(result)

False
True
True


## <a name="expandtabs"></a>7. expandtabs()

__Definition__: returns a copy of string with all tab characters '\t' replaced with whitespace characters until the next multiple of tabsize parameter.  

__Syntax__:  
    
    string.expandtabs(tabsize)  
    
__Parameters__:  
    - tabsize: integer; default size = 8  
__Returns__:  
    - string where all '\t' characters are replaced with whitespace characters until the next multiple of tabsize parameter  

In [32]:
str = 'xyz\t12345\tabc'
print('original str: ',  str)
print()

# no argument is passed; default tabsize is 8
result = str.expandtabs()
print(result)

original str:  xyz	12345	abc

xyz     12345   abc


__How expandtabs() works in Python?__
The expandtabs() method keeps track of the current cursor position.  

The position of first '\t' character in the above program is 3. And, the tabsize is 8 (if argument is not passed).  

The expandtabs() character replaces the '\t' with whitespace until the next tab stop. The position of '\t' is 3 and the first tab stop is 8. Hence, the number of spaces after 'xyz' is 5.  

The next tab stops are the multiples of tabsize. The next tab stops are 16, 24, 32 and so on.  

Now, the position of second '\t' character is 13. And, the next tab stop is 16. Hence, there are 3 spaces after '12345'.

In [33]:
str = "xyz\t12345\tabc"
print('Original String:', str)

# tabsize is set to 2
print('Tabsize 2:', str.expandtabs(2))

# tabsize is set to 3
print('Tabsize 3:', str.expandtabs(3))

# tabsize is set to 4
print('Tabsize 4:', str.expandtabs(4))

# tabsize is set to 5
print('Tabsize 5:', str.expandtabs(5))

# tabsize is set to 6
print('Tabsize 6:', str.expandtabs(6))

Original String: xyz	12345	abc
Tabsize 2: xyz 12345 abc
Tabsize 3: xyz   12345 abc
Tabsize 4: xyz 12345   abc
Tabsize 5: xyz  12345     abc
Tabsize 6: xyz   12345 abc


## <a name="find"></a>8. find()

__Definition__: returns the index of first occurrence of the substring (if found). If not found, it returns -1.  

__Syntax__:  
    
    string.find(sub[start,[ end]])  
    
__Parameters__:  
    - sub: substring to be searched in the original string  
    - start and end (optional): The range str[start:end] within which substring is searched.  
__Returns__:  
    - integer:  
        - substring exists inside the string => it returns the index of the first occurence of the substring.  
        - substring doesn't exist inside the string, it returns -1.  


__Similar:__ [index()](#index) / [rfind()](#rfind)

In [35]:
quote = 'Let it be, let it be, let it be'

# first occurance of 'let it'(case sensitive)
result = quote.find('let it')
print("Substring 'let it':", result)

# find returns -1 if substring not found
result = quote.find('small')
print("Substring 'small ':", result)

# How to use find()
if (quote.find('be,') != -1):
    print("Contains substring 'be,'")
else:
    print("Doesn't contain substring")

Substring 'let it': 11
Substring 'small ': -1
Contains substring 'be,'


In [36]:
quote = 'Do small things with great love'

# Substring is searched in 'hings with great love'
print(quote.find('small things', 10))

# Substring is searched in ' small things with great love' 
print(quote.find('small things', 2))

# Substring is searched in 'hings with great lov'
print(quote.find('o small ', 10, -1))

# Substring is searched in 'll things with'
print(quote.find('things ', 6, 20))

-1
3
-1
9


## <a name="format"></a>9. format()

## <a name="format_map"></a>10. format_map()

## <a name="index"></a>11. index()

__Definition__: returns the index of a substring inside the string. If substring is not found, it raises an exception.  

__Syntax__:  
    
    string.index(sub[start,[ end]])  
    
__Parameters__:  
    - sub: substring to be searched in the original string  
    - start and end (optional): The range str[start:end] within which substring is searched.  
__Returns__:  
    - integer:  
        - substring exists inside the string => it returns the lowest index in the string where substring is found.  
        - substring doesn't exist inside the string => it raises a **ValueError** exception.  


__Similar:__ [find()](#find) / [rfind()](#rfind)

## <a name="isalnum"></a>12. isalnum()

## <a name="isalpha"></a>13. isalpha()

## <a name="isdecimal"></a>14. isdecimal()

## <a name="isdigit"></a>15. isdigit()

## <a name="isidentifier"></a>16. isidentifier()

## <a name="islower"></a>17. islower()

## <a name="isnumeric"></a>18. isnumeric()

## <a name="isprintable"></a>19. isprintable()

## <a name="isspace"></a>20. isspace()

## <a name="istitle"></a>21. istitle()

## <a name="isupper"></a>22. isupper()

## <a name="join"></a>23. join()

## <a name="ljust"></a>24. ljust()

## <a name="lower"></a>25. lower()

__Definition__: converts all uppercase characters in a string into lowercase characters and returns it.   

__Syntax__:  
    
    string.lower()  
    
__Parameters__:  
    - doesn't take any parameter  
__Returns__:  
    - returns the lowercased string from the given string. It converts all uppercase characters to lowercase  
    - If no uppercase characters exist, it returns the original string.  
    - It doesn't modify the original string  
    
__Similar:__ [capitalize()](#capitalize) / [upper()](#upper) / [swapcase()](#swapcase)  

In [18]:
# example string
string = "THIS SHOULD BE LOWERCASE!"
print(string.lower())

# string with numbers
# all alphabets whould be lowercase
string = "Th!s Sh0uLd B3 L0w3rCas3!"
print(string.lower())

this should be lowercase!
th!s sh0uld b3 l0w3rcas3!


In [19]:
# first string
firstString = "PYTHON IS AWESOME!"

# second string
secondString = "PyThOn Is AwEsOmE!"

if(firstString.lower() == secondString.lower()):
    print("The strings are same.")
else:
    print("The strings are not same.")

The strings are same.


## <a name="lstrip"></a>26. lstrip()

## <a name="maketrans"></a>27. maketrans()

## <a name="partition"></a>28. partition()

## <a name="replace"></a>29. replace()

## <a name="rfind"></a>30. rfind()

__Definition__: returns the highest index of the substring (if found). If not found, it returns -1.  

__Syntax__:  
    
    string.rfind(sub[start,[ end]])  
    
__Parameters__:  
    - sub: substring to be searched in the original string  
    - start and end (optional): The range str[start:end] within which substring is searched.  
__Returns__:  
    - integer:  
        - substring exists inside the string => it returns the highest index where substring is found.  
        - substring doesn't exist inside the string, it returns -1.  


__Similar:__ [find()](#find) / [index()](#index)

In [38]:
quote = 'Let it be, let it be, let it be'

result = quote.rfind('let it')
print("Substring 'let it':", result)

result = quote.rfind('small')
print("Substring 'small ':", result)

result = quote.rfind('be,')
if  (result != -1):
  print("Highest index where 'be,' occurs:", result)
else:
  print("Doesn't contain substring")

Substring 'let it': 22
Substring 'small ': -1
Highest index where 'be,' occurs: 18


In [39]:
quote = 'Do small things with great love'

# Substring is searched in 'hings with great love'
print(quote.rfind('things', 10))

# Substring is searched in ' small things with great love' 
print(quote.rfind('t', 2))

# Substring is searched in 'hings with great lov'
print(quote.rfind('o small ', 10, -1))

# Substring is searched in 'll things with'
print(quote.rfind('th', 6, 20))

-1
25
-1
18


## <a name="rindex"></a>31. rindex()

## <a name="rjust"></a>32. rjust()

## <a name="rpartition"></a>33. rpartition()

## <a name="rsplit"></a>34. rsplit()

## <a name="rstrip"></a>35. rstrip()

## <a name="split"></a>36. split()

## <a name="splitlines"></a>37. splitlines()

## <a name="startswith"></a>38. startswith()

__Definition__: returns True if a string starts with the specified suffix. If not, it returns False.  

__Syntax__:  
    
    string.startswith(prefix[start,[ end]])  
    
__Parameters__:  
    - suffix: String or tuple of suffixes to be checked  
    - start (optional): Beginning position where prefix is to be checked within the string  
    - end (optional): Ending position where prefix is to be checked within the string  
__Returns__:  
    - boolean:  
        - True - if string starts with prefix  
        - False - if string doesn't start with prefix  

__Similar:__ [endswith()](#endswith)

In [13]:
text = "Python is easy to learn."

result = text.startswith('is easy')
# returns False
print(result)

result = text.startswith('Python is')
# returns True
print(result)

result = text.startswith('Python is easy to learn.')
# returns True
print(result)

False
True
True


In [14]:
text = "Python programming is easy to learn."

# start parameter: 7
# "programming is easy to learn." string is searched
result = text.startswith('programming is', 7)
print(result)

# Both start and end is provided
# start: 7, end: 26
# "programming is easy" string is searched

result = text.startswith('programming is', 7, 18)
# Returns False
print(result)

result = text.startswith('program', 7, 18)
# returns True
print(result)

True
False
True


### Passing Tuple to startswith()

It's possible to pass a tuple prefixes to the startswith() method in Python.  
If the string ends with any item of the tuple, startswith() returns True.  
If not, it returns False

In [16]:
startswithtext = "programming is easy"
result = startswithtext.startswith(('programming', 'python'))

print(result)

result = startswithtext.startswith(('is', 'easy', 'java'))

print(result)

# With start and end parameter
# 'programming is' string is checked
result = startswithtext.startswith(('programming', 'easy'), 12, 19)

print(result)

True
False
False


## <a name="strip"></a>39. strip()

## <a name="swapcase"></a>40. swapcase()

__Definition__: converts all uppercase characters to lowercase and all lowercase characters to uppercase characters of the given string, and returns it.   

__Syntax__:  
    
    string.swapcase()  
    
__Parameters__:  
    - doesn't take any parameter  
__Returns__:  
    - the string where all uppercase characters are converted to lowercase, and vice-versa  
    - It doesn't modify the original string  

__Similar:__ [capitalize()](#lower) / [lower()](#lower) / [upper()](#upper)  

In [22]:
# example string
string = "THIS SHOULD ALL BE LOWERCASE."
print(string.swapcase())

string = "this should all be uppercase."
print(string.swapcase())

string = "ThIs ShOuLd Be MiXeD cAsEd."
print(string.swapcase())

this should all be lowercase.
THIS SHOULD ALL BE UPPERCASE.
tHiS sHoUlD bE mIxEd CaSeD.


## <a name="title"></a>41. title()

## <a name="translate"></a>42. translate()

## <a name="upper"></a>43. upper()

__Definition__: converts all owercase characters in a string into uppercase characters and returns it.   

__Syntax__:  
    
    string.upper()  
    
__Parameters__:  
    - doesn't take any parameter  
__Returns__:  
    - returns the uppercased string from the given string. It converts all lowercase characters to uppercase.  
    - If no lowercase characters exist, it returns the original string.  
    - It doesn't modify the original string  
    
__Similar:__ [capitalize()](#lower) / [lower()](#lower) / [swapcase()](#swapcase)  

In [20]:
# example string
string = "this should be uppercase!"
print(string.upper())

# string with numbers
# all alphabets whould be lowercase
string = "Th!s Sh0uLd B3 uPp3rCas3!"
print(string.upper())

THIS SHOULD BE UPPERCASE!
TH!S SH0ULD B3 UPP3RCAS3!


In [21]:
# first string
firstString = "python is awesome!"

# second string
secondString = "PyThOn Is AwEsOmE!"

if(firstString.upper() == secondString.upper()):
    print("The strings are same.")
else:
    print("The strings are not same.")

The strings are same.


## <a name="zfill"></a>44. zfill() 